File: backend/app/models/dimensions.py
"""Accounting Dimensions / Cost Centers models."""
import uuid
import enum
from sqlalchemy import (
    Column, String, Text, Numeric, Boolean, Integer,
    ForeignKey, Enum, Date, UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.base import Base, TimestampMixin


# ── Enums ──────────────────────────────────────────────────────────────────────

class DimensionScope(str, enum.Enum):
    FINANCIAL   = "FINANCIAL"
    OPERATIONAL = "OPERATIONAL"
    BOTH        = "BOTH"


class CostCenterType(str, enum.Enum):
    PRODUCTION  = "PRODUCTION"
    WAREHOUSE   = "WAREHOUSE"
    ADMIN       = "ADMIN"
    SALES       = "SALES"
    UTILITIES   = "UTILITIES"
    MAINTENANCE = "MAINTENANCE"
    CORPORATE   = "CORPORATE"
    PROJECT     = "PROJECT"
    OTHER       = "OTHER"


class DimSourceType(str, enum.Enum):
    MANUAL           = "MANUAL"
    DEFAULT          = "DEFAULT"
    INHERITED        = "INHERITED"
    RULE_BASED       = "RULE_BASED"
    SYSTEM_GENERATED = "SYSTEM_GENERATED"


class AllocationBasis(str, enum.Enum):
    FIXED_PCT      = "FIXED_PCT"
    HEADCOUNT      = "HEADCOUNT"
    FLOOR_AREA     = "FLOOR_AREA"
    MACHINE_HOURS  = "MACHINE_HOURS"
    LABOR_HOURS    = "LABOR_HOURS"
    REVENUE        = "REVENUE"
    QTY            = "QTY"
    MANUAL         = "MANUAL"
    CUSTOM_FORMULA = "CUSTOM_FORMULA"


class AllocationFrequency(str, enum.Enum):
    MONTHLY   = "MONTHLY"
    QUARTERLY = "QUARTERLY"
    MANUAL    = "MANUAL"


class AllocationRunStatus(str, enum.Enum):
    DRAFT    = "DRAFT"
    PREVIEWED = "PREVIEWED"
    POSTED   = "POSTED"
    REVERSED = "REVERSED"


class ValidationSeverity(str, enum.Enum):
    WARN  = "WARN"
    BLOCK = "BLOCK"


class DimAIRecStatus(str, enum.Enum):
    PENDING      = "PENDING"
    ACKNOWLEDGED = "ACKNOWLEDGED"
    ACTIONED     = "ACTIONED"
    DISMISSED    = "DISMISSED"


class DimAIAgentType(str, enum.Enum):
    COMPLETENESS_MONITOR = "COMPLETENESS_MONITOR"
    ALLOCATION_OPTIMIZER = "ALLOCATION_OPTIMIZER"
    PROFITABILITY_LENS   = "PROFITABILITY_LENS"


# ── Dimension Type Master ──────────────────────────────────────────────────────

class DimType(Base, TimestampMixin):
    __tablename__ = "dim_types"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    type_code             = Column(String(50), unique=True, nullable=False)
    type_name             = Column(String(120), nullable=False)
    dimension_scope       = Column(Enum(DimensionScope), nullable=False, default=DimensionScope.BOTH)
    hierarchy_enabled     = Column(Boolean, default=True)
    is_mandatory          = Column(Boolean, default=False)
    active                = Column(Boolean, default=True)
    notes                 = Column(Text)

    values = relationship("DimValue", back_populates="dim_type", cascade="all, delete-orphan")
    validation_rules = relationship("DimValidationRule", back_populates="dim_type")


# ── Dimension Value Master ─────────────────────────────────────────────────────

class DimValue(Base, TimestampMixin):
    __tablename__ = "dim_values"
    __table_args__ = (UniqueConstraint("dim_type_id", "dim_code"),)

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    dim_type_id           = Column(UUID(as_uuid=True), ForeignKey("dim_types.id"), nullable=False)
    dim_code              = Column(String(50), nullable=False)
    dim_name              = Column(String(180), nullable=False)
    parent_id             = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=True)
    level_no              = Column(Integer, default=1)
    responsible_user_id   = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    active                = Column(Boolean, default=True)
    start_date            = Column(Date)
    end_date              = Column(Date)
    notes                 = Column(Text)

    dim_type  = relationship("DimType", back_populates="values")
    parent    = relationship("DimValue", remote_side="DimValue.id", foreign_keys=[parent_id])
    children  = relationship("DimValue", foreign_keys=[parent_id])


# ── Cost Center Master ─────────────────────────────────────────────────────────

class CostCenter(Base, TimestampMixin):
    __tablename__ = "cost_centers"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    cost_center_code      = Column(String(50), unique=True, nullable=False)
    cost_center_name      = Column(String(180), nullable=False)
    parent_id             = Column(UUID(as_uuid=True), ForeignKey("cost_centers.id"), nullable=True)
    cost_center_type      = Column(Enum(CostCenterType), nullable=False, default=CostCenterType.ADMIN)
    plant_id              = Column(UUID(as_uuid=True), ForeignKey("warehouses.id"), nullable=True)
    manager_employee_id   = Column(UUID(as_uuid=True), ForeignKey("employees.id"), nullable=True)
    department            = Column(String(100))
    active                = Column(Boolean, default=True)
    start_date            = Column(Date)
    end_date              = Column(Date)
    notes                 = Column(Text)

    parent   = relationship("CostCenter", remote_side="CostCenter.id", foreign_keys=[parent_id])
    children = relationship("CostCenter", foreign_keys=[parent_id])


# ── Transaction Dimension Link ─────────────────────────────────────────────────

class TransactionDimension(Base, TimestampMixin):
    __tablename__ = "transaction_dimensions"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    transaction_type      = Column(String(80), nullable=False)
    transaction_id        = Column(String(80), nullable=False)
    line_id               = Column(String(80))
    dim_type_id           = Column(UUID(as_uuid=True), ForeignKey("dim_types.id"), nullable=False)
    dim_value_id          = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=False)
    source_type           = Column(Enum(DimSourceType), default=DimSourceType.MANUAL)
    locked                = Column(Boolean, default=False)
    notes                 = Column(Text)

    dim_type  = relationship("DimType")
    dim_value = relationship("DimValue")


# ── Dimension Validation Rule ──────────────────────────────────────────────────

class DimValidationRule(Base, TimestampMixin):
    __tablename__ = "dim_validation_rules"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    rule_name             = Column(String(180), nullable=False)
    transaction_type      = Column(String(80))
    gl_account_pattern    = Column(String(100))
    module                = Column(String(80))
    dim_type_id           = Column(UUID(as_uuid=True), ForeignKey("dim_types.id"), nullable=False)
    severity              = Column(Enum(ValidationSeverity), default=ValidationSeverity.WARN)
    active                = Column(Boolean, default=True)
    effective_date        = Column(Date)
    notes                 = Column(Text)

    dim_type = relationship("DimType", back_populates="validation_rules")


# ── Allocation Rule Master ─────────────────────────────────────────────────────

class AllocationRule(Base, TimestampMixin):
    __tablename__ = "allocation_rules"

    id                          = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    rule_code                   = Column(String(50), unique=True, nullable=False)
    rule_name                   = Column(String(180), nullable=False)
    source_dim_type_id          = Column(UUID(as_uuid=True), ForeignKey("dim_types.id"), nullable=False)
    source_dim_value_id         = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=False)
    target_dim_type_id          = Column(UUID(as_uuid=True), ForeignKey("dim_types.id"), nullable=False)
    allocation_basis            = Column(Enum(AllocationBasis), nullable=False, default=AllocationBasis.FIXED_PCT)
    frequency                   = Column(Enum(AllocationFrequency), default=AllocationFrequency.MONTHLY)
    gl_account_cost_pool        = Column(String(50))
    gl_account_allocation_dr    = Column(String(50))
    gl_account_allocation_cr    = Column(String(50))
    active                      = Column(Boolean, default=True)
    notes                       = Column(Text)

    source_dim_type  = relationship("DimType", foreign_keys=[source_dim_type_id])
    source_dim_value = relationship("DimValue", foreign_keys=[source_dim_value_id])
    target_dim_type  = relationship("DimType", foreign_keys=[target_dim_type_id])
    lines            = relationship("AllocationRuleLine", back_populates="rule", cascade="all, delete-orphan")
    runs             = relationship("AllocationRun", back_populates="rule")


# ── Allocation Rule Lines ──────────────────────────────────────────────────────

class AllocationRuleLine(Base, TimestampMixin):
    __tablename__ = "allocation_rule_lines"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    rule_id               = Column(UUID(as_uuid=True), ForeignKey("allocation_rules.id"), nullable=False)
    target_dim_value_id   = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=False)
    fixed_pct             = Column(Numeric(8, 4))
    weight_value          = Column(Numeric(18, 4))
    active                = Column(Boolean, default=True)
    notes                 = Column(Text)

    rule             = relationship("AllocationRule", back_populates="lines")
    target_dim_value = relationship("DimValue")


# ── Allocation Run ─────────────────────────────────────────────────────────────

class AllocationRun(Base, TimestampMixin):
    __tablename__ = "allocation_runs"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    rule_id               = Column(UUID(as_uuid=True), ForeignKey("allocation_rules.id"), nullable=False)
    period_start          = Column(Date, nullable=False)
    period_end            = Column(Date, nullable=False)
    dry_run               = Column(Boolean, default=True)
    status                = Column(Enum(AllocationRunStatus), default=AllocationRunStatus.DRAFT)
    source_pool_amount    = Column(Numeric(18, 4), default=0)
    total_allocated       = Column(Numeric(18, 4), default=0)
    run_notes             = Column(Text)
    posted_by_id          = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)

    rule      = relationship("AllocationRule", back_populates="runs")
    lines     = relationship("AllocationRunLine", back_populates="run", cascade="all, delete-orphan")


# ── Allocation Run Lines ───────────────────────────────────────────────────────

class AllocationRunLine(Base, TimestampMixin):
    __tablename__ = "allocation_run_lines"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    run_id                = Column(UUID(as_uuid=True), ForeignKey("allocation_runs.id"), nullable=False)
    target_dim_value_id   = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=False)
    pct_applied           = Column(Numeric(8, 4))
    allocated_amount      = Column(Numeric(18, 4))
    journal_entry_id      = Column(String(80))

    run              = relationship("AllocationRun", back_populates="lines")
    target_dim_value = relationship("DimValue")


# ── Dimension Default Rule ─────────────────────────────────────────────────────

class DimDefaultRule(Base, TimestampMixin):
    __tablename__ = "dim_default_rules"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    rule_name             = Column(String(180), nullable=False)
    transaction_type      = Column(String(80), nullable=False)
    source_field          = Column(String(100))
    source_field_value    = Column(String(200))
    dim_type_id           = Column(UUID(as_uuid=True), ForeignKey("dim_types.id"), nullable=False)
    dim_value_id          = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=False)
    priority              = Column(Integer, default=10)
    active                = Column(Boolean, default=True)

    dim_type  = relationship("DimType")
    dim_value = relationship("DimValue")


# ── Reclassification Record ────────────────────────────────────────────────────

class DimReclassification(Base, TimestampMixin):
    __tablename__ = "dim_reclassifications"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    transaction_type      = Column(String(80), nullable=False)
    transaction_id        = Column(String(80), nullable=False)
    dim_type_id           = Column(UUID(as_uuid=True), ForeignKey("dim_types.id"), nullable=False)
    old_dim_value_id      = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=True)
    new_dim_value_id      = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=False)
    reason                = Column(Text, nullable=False)
    journal_entry_ref     = Column(String(80))
    reclassified_by_id    = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)

    dim_type      = relationship("DimType")
    old_dim_value = relationship("DimValue", foreign_keys=[old_dim_value_id])
    new_dim_value = relationship("DimValue", foreign_keys=[new_dim_value_id])


# ── AI Recommendations ────────────────────────────────────────────────────────

class DimAIRecommendation(Base, TimestampMixin):
    __tablename__ = "dim_ai_recommendations"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_type            = Column(Enum(DimAIAgentType), nullable=False)
    title                 = Column(String(300), nullable=False)
    detail                = Column(Text)
    severity              = Column(String(20), default="info")
    status                = Column(Enum(DimAIRecStatus), default=DimAIRecStatus.PENDING)
    dim_type_id           = Column(UUID(as_uuid=True), ForeignKey("dim_types.id"), nullable=True)
    dim_value_id          = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=True)
    allocation_rule_id    = Column(UUID(as_uuid=True), ForeignKey("allocation_rules.id"), nullable=True)

File: frontend/src/lib/dimensions.ts
import axios from "axios";

const API = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000/api/v1";
const r = <T>(p: Promise<{ data: T }>) => p.then((res) => res.data);
const ax = axios.create({ baseURL: API });

// ── Enums & Types ──────────────────────────────────────────────────────────────

export type DimensionScope = "FINANCIAL" | "OPERATIONAL" | "BOTH";
export type CostCenterType =
  | "PRODUCTION" | "WAREHOUSE" | "ADMIN" | "SALES"
  | "UTILITIES" | "MAINTENANCE" | "CORPORATE" | "PROJECT" | "OTHER";
export type DimSourceType = "MANUAL" | "DEFAULT" | "INHERITED" | "RULE_BASED" | "SYSTEM_GENERATED";
export type AllocationBasis =
  | "FIXED_PCT" | "HEADCOUNT" | "FLOOR_AREA" | "MACHINE_HOURS"
  | "LABOR_HOURS" | "REVENUE" | "QTY" | "MANUAL" | "CUSTOM_FORMULA";
export type AllocationFrequency = "MONTHLY" | "QUARTERLY" | "MANUAL";
export type AllocationRunStatus = "DRAFT" | "PREVIEWED" | "POSTED" | "REVERSED";
export type ValidationSeverity = "WARN" | "BLOCK";
export type DimAIAgentType = "COMPLETENESS_MONITOR" | "ALLOCATION_OPTIMIZER" | "PROFITABILITY_LENS";
export type DimAIRecStatus = "PENDING" | "ACKNOWLEDGED" | "ACTIONED" | "DISMISSED";

// ── Dimension Type ─────────────────────────────────────────────────────────────

export interface DimType {
  id: string;
  type_code: string;
  type_name: string;
  dimension_scope: DimensionScope;
  hierarchy_enabled: boolean;
  is_mandatory: boolean;
  active: boolean;
  notes?: string;
  value_count: number;
}

// ── Dimension Value ────────────────────────────────────────────────────────────

export interface DimValue {
  id: string;
  dim_type_id: string;
  dim_code: string;
  dim_name: string;
  parent_id?: string;
  level_no: number;
  active: boolean;
  start_date?: string;
  end_date?: string;
  notes?: string;
  dim_type_name?: string;
  parent_name?: string;
  children_count: number;
}

// ── Cost Center ────────────────────────────────────────────────────────────────

export interface CostCenter {
  id: string;
  cost_center_code: string;
  cost_center_name: string;
  parent_id?: string;
  cost_center_type: CostCenterType;
  department?: string;
  active: boolean;
  start_date?: string;
  end_date?: string;
  notes?: string;
  parent_name?: string;
  children_count: number;
}

// ── Transaction Dimension Tag ──────────────────────────────────────────────────

export interface TransactionDimTag {
  id: string;
  transaction_type: string;
  transaction_id: string;
  line_id?: string;
  dim_type_id: string;
  dim_value_id: string;
  source_type: DimSourceType;
  locked: boolean;
  dim_type_name?: string;
  dim_value_name?: string;
}

// ── Validation Rule ────────────────────────────────────────────────────────────

export interface DimValidationRule {
  id: string;
  rule_name: string;
  transaction_type?: string;
  gl_account_pattern?: string;
  module?: string;
  dim_type_id: string;
  severity: ValidationSeverity;
  active: boolean;
  dim_type_name?: string;
}

// ── Default Rule ───────────────────────────────────────────────────────────────

export interface DimDefaultRule {
  id: string;
  rule_name: string;
  transaction_type: string;
  source_field?: string;
  source_field_value?: string;
  dim_type_id: string;
  dim_value_id: string;
  priority: number;
  active: boolean;
  dim_type_name?: string;
  dim_value_name?: string;
}

// ── Allocation Rule ────────────────────────────────────────────────────────────

export interface AllocationRuleLine {
  id: string;
  target_dim_value_id: string;
  target_dim_value_name?: string;
  fixed_pct?: number;
  weight_value?: number;
  active: boolean;
}

export interface AllocationRule {
  id: string;
  rule_code: string;
  rule_name: string;
  source_dim_type_id: string;
  source_dim_value_id: string;
  target_dim_type_id: string;
  allocation_basis: AllocationBasis;
  frequency: AllocationFrequency;
  gl_account_cost_pool?: string;
  active: boolean;
  source_dim_value_name?: string;
  target_dim_type_name?: string;
  lines: AllocationRuleLine[];
}

// ── Allocation Run ─────────────────────────────────────────────────────────────

export interface AllocationRunLine {
  id: string;
  target_dim_value_id: string;
  target_dim_value_name?: string;
  pct_applied?: number;
  allocated_amount?: number;
  journal_entry_id?: string;
}

export interface AllocationRun {
  id: string;
  rule_id: string;
  rule_name?: string;
  period_start: string;
  period_end: string;
  dry_run: boolean;
  status: AllocationRunStatus;
  source_pool_amount: number;
  total_allocated: number;
  run_notes?: string;
  lines: AllocationRunLine[];
}

// ── Reclassification ───────────────────────────────────────────────────────────

export interface Reclassification {
  id: string;
  transaction_type: string;
  transaction_id: string;
  dim_type_id: string;
  dim_type_name?: string;
  old_dim_value_id?: string;
  old_dim_value_name?: string;
  new_dim_value_id: string;
  new_dim_value_name?: string;
  reason: string;
  journal_entry_ref?: string;
  created_at?: string;
}

// ── AI Recommendations ─────────────────────────────────────────────────────────

export interface DimAIRec {
  id: string;
  agent_type: DimAIAgentType;
  title: string;
  detail?: string;
  severity: string;
  status: DimAIRecStatus;
}

// ── Dashboard ──────────────────────────────────────────────────────────────────

export interface DimDashboard {
  total_dim_types: number;
  total_dim_values: number;
  total_cost_centers: number;
  active_allocation_rules: number;
  untagged_transactions_today: number;
  pending_ai_recs: number;
}

// ── Labels ─────────────────────────────────────────────────────────────────────

export const CC_TYPE_LABEL: Record<CostCenterType, string> = {
  PRODUCTION:  "Production",
  WAREHOUSE:   "Warehouse",
  ADMIN:       "Administration",
  SALES:       "Sales",
  UTILITIES:   "Utilities",
  MAINTENANCE: "Maintenance",
  CORPORATE:   "Corporate",
  PROJECT:     "Project",
  OTHER:       "Other",
};

export const ALLOC_BASIS_LABEL: Record<AllocationBasis, string> = {
  FIXED_PCT:      "Fixed %",
  HEADCOUNT:      "Headcount",
  FLOOR_AREA:     "Floor Area",
  MACHINE_HOURS:  "Machine Hours",
  LABOR_HOURS:    "Labor Hours",
  REVENUE:        "Revenue",
  QTY:            "Quantity",
  MANUAL:         "Manual",
  CUSTOM_FORMULA: "Custom Formula",
};

export const AGENT_LABEL: Record<DimAIAgentType, string> = {
  COMPLETENESS_MONITOR: "Completeness Monitor",
  ALLOCATION_OPTIMIZER: "Allocation Optimizer",
  PROFITABILITY_LENS:   "Profitability Lens",
};

export const SEVERITY_BADGE: Record<string, string> = {
  info:     "bg-blue-100 text-blue-700",
  warning:  "bg-yellow-100 text-yellow-700",
  critical: "bg-red-100 text-red-700",
};

// ── API Client ─────────────────────────────────────────────────────────────────

export const dimApi = {
  // Dashboard
  getDashboard: () => r<DimDashboard>(ax.get("/dimensions/dashboard")),

  // Dimension Types
  getTypes: (activeOnly?: boolean) =>
    r<DimType[]>(ax.get("/dimensions/types", { params: { active_only: activeOnly } })),
  createType: (data: Partial<DimType>) =>
    r<DimType>(ax.post("/dimensions/types", data)),
  updateType: (id: string, data: Partial<DimType>) =>
    r<DimType>(ax.patch(`/dimensions/types/${id}`, data)),

  // Dimension Values
  getValues: (dimTypeId?: string, activeOnly?: boolean) =>
    r<DimValue[]>(ax.get("/dimensions/values", { params: { dim_type_id: dimTypeId, active_only: activeOnly } })),
  createValue: (data: Partial<DimValue>) =>
    r<DimValue>(ax.post("/dimensions/values", data)),

  // Cost Centers
  getCostCenters: (activeOnly?: boolean) =>
    r<CostCenter[]>(ax.get("/dimensions/cost-centers", { params: { active_only: activeOnly } })),
  createCostCenter: (data: Partial<CostCenter>) =>
    r<CostCenter>(ax.post("/dimensions/cost-centers", data)),
  updateCostCenter: (id: string, data: Partial<CostCenter>) =>
    r<CostCenter>(ax.patch(`/dimensions/cost-centers/${id}`, data)),

  // Tags
  getTags: (transactionType: string, transactionId: string) =>
    r<TransactionDimTag[]>(ax.get("/dimensions/tags", { params: { transaction_type: transactionType, transaction_id: transactionId } })),
  tagTransaction: (data: Partial<TransactionDimTag>) =>
    r<TransactionDimTag>(ax.post("/dimensions/tags", data)),

  // Validation Rules
  getValidationRules: () => r<DimValidationRule[]>(ax.get("/dimensions/validation-rules")),
  createValidationRule: (data: Partial<DimValidationRule>) =>
    r<DimValidationRule>(ax.post("/dimensions/validation-rules", data)),

  // Default Rules
  getDefaultRules: () => r<DimDefaultRule[]>(ax.get("/dimensions/default-rules")),
  createDefaultRule: (data: Partial<DimDefaultRule>) =>
    r<DimDefaultRule>(ax.post("/dimensions/default-rules", data)),

  // Allocation Rules
  getAllocationRules: () => r<AllocationRule[]>(ax.get("/dimensions/allocation-rules")),
  createAllocationRule: (data: Partial<AllocationRule>) =>
    r<AllocationRule>(ax.post("/dimensions/allocation-rules", data)),

  // Allocation Runs
  getAllocationRuns: (ruleId?: string) =>
    r<AllocationRun[]>(ax.get("/dimensions/allocation-runs", { params: { rule_id: ruleId } })),
  previewAllocation: (data: object) =>
    r<AllocationRun>(ax.post("/dimensions/allocation-runs/preview", data)),
  postAllocation: (data: object) =>
    r<AllocationRun>(ax.post("/dimensions/allocation-runs/post", data)),

  // Reclassification
  getReclassifications: (transactionType?: string) =>
    r<Reclassification[]>(ax.get("/dimensions/reclassifications", { params: { transaction_type: transactionType } })),
  reclassify: (data: object) =>
    r<Reclassification>(ax.post("/dimensions/reclassify", data)),

  // Reports
  getTaggingCompleteness: () =>
    r<any[]>(ax.get("/dimensions/reports/tagging-completeness")),

  // AI
  runAgents: () => r<{ generated: number }>(ax.post("/dimensions/ai/run-agents")),
  getAIRecs: (status?: DimAIRecStatus) =>
    r<DimAIRec[]>(ax.get("/dimensions/ai/recommendations", { params: { status } })),
  ackAIRec: (id: string, data: { status: DimAIRecStatus }) =>
    r<DimAIRec>(ax.patch(`/dimensions/ai/recommendations/${id}`, data)),
};
